hawc/tools/tables/sot.py

```diff
@@ -0,0 +1,140 @@
+
+import pandas as pd
+from pydantic import BaseModel, Field, conint
+
+from hawc.apps.animal.models import Endpoint
+from hawc.apps.materialized.models import FinalRiskOfBiasScore
+from hawc.apps.riskofbias.constants import SCORE_SHADES, SCORE_SYMBOLS
+
+from .base import BaseCell, BaseCellGroup, BaseTable
+from .generic import GenericCell
+from .parser import QuillParser, color_background, tag_wrapper
+
+
+class AttributeChoices(Enum):
+    RobScore = "rob_score"
+    StudyName = "study_name"
+    SpeciesStrainSex = "species_strain_sex"
+    Doses = "doses"
+    FreeHtml = "free_html"
+
+
+class Subheader(BaseModel):
+    label: str
+    start: conint(ge=0)
+    length: conint(ge=1)
+
+
+class Column(BaseModel):
+    label: str
+    attribute: AttributeChoices
+    metric: Optional[int]
+    key: str
+
+
+class Row(BaseModel):
+    id: int
+    type: str
+    customized: List
+
+
+class JudgementCell(BaseCell):
+    judgement: int
+
+    def to_docx(self, parser: QuillParser, block):
+        text = SCORE_SYMBOLS[self.judgement]
+        color = SCORE_SHADES[self.judgement].strip("#")
+        color_background(block, color)
+        return parser.feed(tag_wrapper(text, "p"), block)
+
+
+class StudyOutcomeTable(BaseTable):
+    assessment_id: int
+    subheaders: List[Subheader]
+    cell_columns: List[Column] = Field([], alias="columns")
+    cell_rows: List[Row] = Field([], alias="rows")
+
+    def get_html(self, data, row, attribute):
+        selection = data.loc[data["study id"] == row.id]
+        if attribute == "study_name":
+            text = selection["study citation"].values[0]
+            return tag_wrapper(text, "p")
+        elif attribute == "species_strain_sex":
+            text = selection["species"].values[0]
+            return tag_wrapper(text, "p")
+        elif attribute == "doses":
+            text = selection["dose units"].values[0]
+            return tag_wrapper(text, "p")
+        elif attribute == "free_html":
+            return tag_wrapper("", "p")
+
+    def _subheaders_group(self):
+        cells = []
+        for subheader in self.subheaders:
+            html = tag_wrapper(subheader.label, "p", "strong")
+            cells.append(
+                GenericCell.parse_args(True, 0, subheader.start, 1, subheader.length, html)
+            )
+        return BaseCellGroup.construct(cells=cells)
+
+    def _columns_group(self):
+        cells = []
+        for i, col in enumerate(self.cell_columns):
+            html = tag_wrapper(col.label, "p", "strong")
+            cells.append(GenericCell.parse_args(True, 0, i, 1, 1, html))
+        return BaseCellGroup.construct(cells=cells)
+
+    def _rows_group(self):
+        cells = []
+
+        study_ids = {row.id for row in self.cell_rows if row.type == "study"}
+        final_scores = pd.DataFrame.from_records(
+            FinalRiskOfBiasScore.objects.filter(study_id__in=study_ids).values()
+        )
+        bioassay_data = Endpoint.heatmap_study_df(
```

Comment:
I am using `heatmap_study_df` for all of the non-RoB data. However, it may be a little incomplete when it comes to columns we want to show, and things like sex/species/strains are aggregated by piping them together, which may not be what we want for this table. Should we:
1) split strings that are piped together and present it in more of a readable format for the table
2) add other fields to `heatmap_study_df` that we need for this table (ie it provides dose units, but not actual doses)
3) OR just make a new API endpoint altogether specifically suited to this table (maybe even revisit the animal bioassay materialized view?)

Reply:
I was thinking we could use the `heatmap_doses_df` actually so we could get down to loel/noel data in the future.  We'd need to have rollup functions by data type, so for now, you could do a rollup for species by study for example as a starting point and